webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/domain/nodes/Properties.scala

```diff
@@ -194,24 +269,77 @@ object GenericProperty {
     }
   }
 
+
   /*
    * Merge two json values, overriding or merging recursively
    */
-  def mergeValues(oldValue: ConfigValue, newValue: ConfigValue): ConfigValue = {
-    newValue.withFallback(oldValue)
+  def mergeValues(oldValue: ConfigValue, newValue: ConfigValue, mode: InheritMode): ConfigValue = {
+    import ConfigValueType._
+    import InheritMode._
+    import java.util.{List => juList, Map => juMap}
+    import scala.jdk.CollectionConverters._
+
+    def stringPlus(a: ConfigValue, b: ConfigValue): ConfigValue = {
+      ConfigValueFactory.fromAnyRef(a.unwrapped().asInstanceOf[String] ++ b.unwrapped().asInstanceOf[String])
+    }
+    def listPlus(a: ConfigValue, b: ConfigValue): ConfigValue = {
+        val l = a.unwrapped().asInstanceOf[juList[Any]]
+        l.addAll(b.unwrapped().asInstanceOf[juList[Any]])
+        ConfigValueFactory.fromIterable(l)
+    }
+    def objectPlus(a: ConfigValue, b: ConfigValue, m: InheritMode): ConfigValue = {
+        val o = a.unwrapped().asInstanceOf[juMap[String, Any]].asScala
+        val n = b.unwrapped().asInstanceOf[juMap[String, Any]].asScala
+        val mergeNew = n.map { case(k,vn) => o.get(k) match {
+          case None     => (k, vn)
+          case Some(vo) => (k, mergeValues(ConfigValueFactory.fromAnyRef(vo), ConfigValueFactory.fromAnyRef(vn), m))
+        } }
+        ConfigValueFactory.fromMap((o ++ mergeNew).asJava)
+    }
+
+    (oldValue.valueType(), newValue.valueType()) match {
+      case (STRING,STRING) =>
+        mode.forString match {
+          case StringMode.Override => newValue
+          case StringMode.Prepend  => stringPlus(newValue, oldValue)
+          case StringMode.Append   => stringPlus(oldValue, newValue)
+        }
+      case (_,NULL|STRING|NUMBER|BOOLEAN) => newValue //override in all case
+      case (LIST, LIST) => //override by default
+        mode.forArray match {
+          case ArrayMode.Override => newValue
+          case ArrayMode.Prepend  => listPlus(newValue, oldValue)
+          case ArrayMode.Append   => listPlus(oldValue, newValue)
+        }
+      case (_, LIST) => newValue  //override in all case
+      case (OBJECT,OBJECT) => //merge by default
+        mode.forObject match {
+          case ObjectMode.Override => newValue
+          case ObjectMode.Merge    => objectPlus(oldValue, newValue, mode)
+        }
+      case (_,OBJECT) => newValue //override
+    }
   }
 
   /**
    * Merge two properties. newProp values will win.
    * You should have check before that "name" and "provider" are OK.
    */
   def mergeConfig(oldProp: Config, newProp: Config): Config = {
-    newProp.withFallback(oldProp)
+    val mode = (GenericProperty.getMode(oldProp), GenericProperty.getMode(newProp)) match {
+      case (None, None)                                   => None
+      case (None, Some(mode))                             => Some(mode)
+      case (Some(mode), None | Some(InheritMode.Default)) => Some(mode)
```

Comment:
You should not get oldProperty mode if someone has selected the same mode than default

Reply:
Hum. Yes, you are right, it's a sequel of the past when it was not in option!